### PR TITLE
PUBDEV-6176: moved StackedEnsembleModel to hex.ensemble package 

### PR DESCRIPTION
**Pull request by seb-h2o**
https://0xdata.atlassian.net/browse/PUBDEV-6176

This implies changes in how SE scoring is done in predictScoreImpl, especially deactivated computeMetrics for base models at it appeared useless (we only need predictions).

Also fixes https://0xdata.atlassian.net/browse/PUBDEV-6266

Adding migration document for those changes:
This impacts only Java API:
- `hex.StackedEnsembleModel` -> `hex.ensemble.StackedEnsembleModel`
- `hex.StackedEnsembleModel.MetalearnerAlgorithm` -> `hex.ensemble.Metalearner.Algorithm`
- removed public access to following internal methods:
  - `StackedEnsembleModel.checkAndInheritModelProperties`.
  - `StackedEnsembleModel.doScoreOrCopyMetrics`.
  - `StackedEnsembleModel.doScoreMetricsOneFrame`.
  - `StackedEnsemble.addModelPredictionsToLevelOneFrame`.

### CODE REVIEW

File: h2o-algos/src/main/java/hex/ensemble/StackedEnsembleModel.java [hunk 1]
```diff
@@ -135,44 +119,22 @@ public void initMetalearnerParams(MetalearnerAlgorithm algo) {
    */
   @Override
   protected Frame predictScoreImpl(Frame fr, Frame adaptFrm, String destination_key, Job j, boolean computeMetrics, CFuncRef customMetricFunc) {
```
Comment (seb-h2o) [inline h2o-algos/src/main/java/hex/ensemble/StackedEnsembleModel.java]: @michalkurka : please review the rewrite of this `predictScoreImpl` method for Stacked Ensemble, that's also what fixes the issue PUBDEV-6266 (cf. test below).

File: h2o-algos/src/test/java/hex/ensemble/StackedEnsembleTest.java [hunk 2]
```diff
@@ -417,6 +415,73 @@ public void testBaseModelPredictionsCaching() {
       }
     }
 
+    
+    @Test
+    public void test_SE_with_GLM_can_do_predictions_on_frames_with_unseen_categorical_values() {
```
Comment (seb-h2o) [inline h2o-algos/src/test/java/hex/ensemble/StackedEnsembleTest.java]: tested that this test fails on current master.

File: h2o-docs/src/product/migrating.rst [hunk 3]
```diff
@@ -1,4 +1,30 @@
 .. _migration:
+.. include:: <isonum.txt>
+
+Migration Guide
+###############
+
+Migrating between H2O-3 versions
```
Comment (seb-h2o) [inline h2o-docs/src/product/migrating.rst]: @angela0xdata : I introduced this new section, and put it on top of the file as I believe that there are way more people today likely to migrate between h2o-3 versions than from h20-2 to 3.
Of course, any suggestion welcome.

Comment (angela0xdata) [reply h2o-docs/src/product/migrating.rst]: @seb-h2o I don't understand the include statement. What is <isonum.txt>?

File: h2o-docs/src/product/migrating.rst [hunk 4]
```diff
@@ -1,4 +1,30 @@
...
+--------------------------
+
+Java API
+~~~~~~~~
+
+The following classes were moved and/or renamed:
+
+- ``hex.StackedEnsembleModel`` |rarr| ``hex.ensemble.StackedEnsembleModel``.
+- ``hex.StackedEnsembleModel.MetalearnerAlgorithm`` |rarr| ``hex.ensemble.Metalearner.Algorithm``.
+
```
Comment (angela0xdata) [inline h2o-docs/src/product/migrating.rst]: What is |rarr| in the above two bullets? These references are causing an error when building the docs.

**Comment by angela0xdata** [comment]
@seb-h2o I made some updates to the heading levels. Also, please see comments regarding references and the include statement in the doc.

File: h2o-docs/src/product/migrating.rst (hunk 4 above)
Comment (seb-h2o) [reply h2o-docs/src/product/migrating.rst]: @angela0xdata those are non-ascii characters (right-arrow in this case), hence the `isonum` import above.
I built the docs locally before and it worked fine, that's why I included it... maybe it depends on the environment though. If it doesn't work everywhere, I'll remove it.

Comment (seb-h2o) [reply h2o-docs/src/product/migrating.rst]: for reference: http://docutils.sourceforge.net/docs/ref/rst/definitions.html

**Comment by seb-h2o** [comment]
@angela0xdata I removed the import for special characters and using a table formatting instead: feel free to try building the docs again.
Also thanks for the other changes! btw, don't know if you did the header changes mainly for consistency, but FYI rST format doesn't formally require to change all headers in this case, as it's rebuilding the hierarchy when parsing the file (matching similar underlines): it's a cool feature indeed as it allows shifting entire sub-trees just by adding/removing one single header and without having to modify the existing... pretty neat.

**Comment by angela0xdata** [comment]
@seb-h2o Thanks for the info. Everything does build locally, but the build produced additional errors (which aren't harmful; I just don't like to see them if I can avoid them). But I agree that rst can be pretty nifty.

**Comment by seb-h2o** [comment]
@michalkurka would appreciate if you could find time to review this today: tests are fixed, and this ticket is blocking me to complete https://github.com/h2oai/h2o-3/pull/3273 (leaking keys in UT after scoring, fixed by this scoring rewrite).

File: h2o-docs/src/product/migrating.rst [hunk 5]
```diff
@@ -1,9 +1,39 @@
...
+Java API
+''''''''
+
+The following classes were moved and/or renamed:
+
+=================================================   ======================================
+  Until 3.22                                          From 3.24
+=================================================   ======================================
+``hex.StackedEnsembleModel``                        ``hex.ensemble.StackedEnsembleModel``
```
Comment (michalkurka) [inline h2o-docs/src/product/migrating.rst]: @jakubhava how does this affect Sparkling Water?

File: h2o-algos/src/main/java/hex/ensemble/StackedEnsembleModel.java (hunk 1 above)
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/ensemble/StackedEnsembleModel.java]: Looks good to me so far - I will check one more time

File: h2o-core/src/main/java/water/util/ArrayUtils.java [hunk 6]
```diff
@@ -1437,7 +1437,7 @@ public int compare(Integer x, Integer y) {
     if(s == null)return ary;
     int cnt = 0;
     int idx = find(ary,s);
-    while(idx > 0) {
+    while(idx >= 0) {
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/util/ArrayUtils.java]: wow!

File: h2o-docs/src/product/migrating.rst [hunk 7]
```diff
@@ -1,9 +1,39 @@
...
+``hex.StackedEnsembleModel.MetalearnerAlgorithm``   ``hex.ensemble.Metalearner.Algorithm``
+``ai.h2o.automl.AutoML.algo``                       ``ai.h2o.automl.Algo``
+=================================================   ======================================
+
+Some internal methods of ``StackedEnsemble`` and ``StackedEnsembleModel`` are no longer public, but this should not impact anyone.
+
+Migrating to H2O-3
```
Comment (michalkurka) [inline h2o-docs/src/product/migrating.rst]: @angela0xdata can we remove the h2o-2 to h2o-3 section, it is really not relevant anymore

File: h2o-docs/src/product/migrating.rst (hunk 5 above)
Comment (jakubhava) [reply h2o-docs/src/product/migrating.rst]: nope, we're good there 👍 

Thanks for the heads up though!

Comment (michalkurka) [reply h2o-docs/src/product/migrating.rst]: How about the other 2? I remember last time we renamed `algos` it caused SW tests to fail.

Comment (jakubhava) [reply h2o-docs/src/product/migrating.rst]: the answer applies to other ones as well, the `algo->Algo` change is already in a few last SW releases, green. @seb-h2o did that in a smart backwards-compatible way.

**Comment by angela0xdata** [comment]
@michalkurka I'm fine removing that legacy migrating section, but I would also defer to @tomkraljevic.

**Comment by angela0xdata** [comment]
@seb-h2o @michalkurka @jakubhava, Tom has said agreed that we can remove that "Migrating to H2O 3" section.

File: h2o-docs/src/product/migrating.rst (hunk 5 above)
Comment (michalkurka) [reply h2o-docs/src/product/migrating.rst]: perfect, thanks!

**Comment by seb-h2o** [comment]
@angela0xdata merging now, will remove the migration section on a separate tiny PR

**Pull request merged**